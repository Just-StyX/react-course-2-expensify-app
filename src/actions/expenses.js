import { v1 as uuid } from "uuid";
import database from '../firebase/firebase'
import { ref, set, remove, update, onValue, off, push, onChildRemoved, onChildChanged, onChildAdded } from 'firebase/database'

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt }
        const reference = ref(database, 'expenses')
        return push(reference, expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))})
    }
}

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        const reference = ref(database, `expenses/${id}`)
        return remove(reference).then(() => {
            dispatch(removeExpense({ id }))
        })
    }
}

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id, 
    updates
})

export const setExpense = (expenses) => ({
    type: 'SET_EXPENSE',
    expenses
})

export const startSetExpenses = () => {
    return (dispatch) => {
        const reference = ref(database, 'expenses')
        return onValue(reference, (snapshot) => {
            const expenses = [];

            snapshot.forEach((child) => {
                expenses.push({
                    id: child.key,
                    ...child.val()
                })
            })
            dispatch(setExpense(expenses))
        })
    }
}