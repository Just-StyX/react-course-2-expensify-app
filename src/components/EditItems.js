import React from 'react'
import { NavLink } from 'react-router-dom'

const EditItems = () => (
    <div>
        <p>Edit the expense items <NavLink to="edit" style={({ isActive }) => ({ color: isActive ? 'green' : 'blue', fontWeight: isActive ? 'bold': "" })}>here</NavLink></p>
    </div>
)

export default EditItems