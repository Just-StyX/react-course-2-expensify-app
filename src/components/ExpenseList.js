import React from "react";
import { connect } from "react-redux";
//import numeral from "numeral";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
  return (
    <div className="content-container">
      <div className="lists-header">
        <div className="show-mobile">Expenses</div>
        <div className="show-desktop">Expense</div>
        <div className="show-desktop">Amount</div>
      </div>
        <div className="lists-body">
        {props.expenses.length === 0 ?(<div className="lists-item lists-item--message"><span>No Expenses</span></div>) : (props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        }))}
        </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
