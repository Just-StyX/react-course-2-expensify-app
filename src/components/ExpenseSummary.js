import React from 'react'
import numeral from 'numeral'
import { connect } from 'react-redux'
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from '../selectors/expense-total'

export const ExpenseSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
    return(
        <div>
            <h3>You are viewimg {expenseCount} {expenseWord} whose total amount is {formattedExpensesTotal}</h3>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
      expenseCount: visibleExpenses.length,
      expensesTotal: selectExpensesTotal(visibleExpenses) 
    };
  };
  
  export default connect(mapStateToProps)(ExpenseSummary);