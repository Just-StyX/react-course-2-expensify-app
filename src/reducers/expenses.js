const ADD_EXPENSE = "ADD_EXPENSE";
const REMOVE_EXPENSE = "REMOVE_EXPENSE";
const EDIT_EXPENSE = "EDIT_EXPENSE";
const SET_EXPENSE = "SET_EXPENSE"

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [...state, action.expense];
    case REMOVE_EXPENSE:
      return state.filter(({ id }) => id !== action.id);
    case EDIT_EXPENSE:
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    case SET_EXPENSE:
      return action.expenses
    default:
      return state;
  }
};

export default expensesReducer;
