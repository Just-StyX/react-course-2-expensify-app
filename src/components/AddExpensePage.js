import React from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

class AddExpensePageClass extends React.Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense);
    this.props.history("/dashboard");
  };
  render() {
    const { history } = this.props;
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch(startAddExpense(expense))
});

export const AddExpensePage = (props) => {
  const history = useNavigate();
  return <AddExpensePageClass {...props}  history={history}/>;
};


export default connect(undefined, mapDispatchToProps)(AddExpensePage);
