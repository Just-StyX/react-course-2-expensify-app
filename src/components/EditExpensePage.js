import React from "react";
import { useNavigate } from "react-router-dom";
import { connect, useDispatch } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { setEditExpense, startRemoveExpense } from "../actions/expenses";

class EditExpensePageClass extends React.Component {
  onSubmit = (expense) => {
    this.props.dispatch(setEditExpense(this.props.expense.id, expense));
    this.props.history("/dashboard");
  };
  onRemove = () => {
    this.props.dispatch(startRemoveExpense({ id: this.props.expense.id }));
    this.props.history("/dashboard");
  };
  render() {
    const { history } = this.props;
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

export const EditExpensePage = (props) => {
  const history = useNavigate();
  return <EditExpensePageClass {...props} history={history} />;
};

const mapStateToProps = (state, ownProps) => {
  const pathArray = window.location.pathname.split("/");
  const id = pathArray[2];
  return {
    expense: state.expenses.find((expense) => expense.id === id)
  };
};

export default connect(mapStateToProps)(EditExpensePage);
