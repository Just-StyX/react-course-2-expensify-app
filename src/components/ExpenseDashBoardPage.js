import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashBoardPage = () => {
  return (
    <div>
      <ExpenseListFilters />
      <ExpenseList />
      <p>This is from my dashboard component</p>
    </div>
  );
};

export default ExpenseDashBoardPage;
