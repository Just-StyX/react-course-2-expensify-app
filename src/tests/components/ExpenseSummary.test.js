import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses'

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235}/>)
    expect(toJSON(wrapper)).toMatchSnapshot()
}) 

test('should correctly render ExpensesSummary with multiple expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={23} expensesTotal={235468986670}/>)
    expect(toJSON(wrapper)).toMatchSnapshot()
}) 