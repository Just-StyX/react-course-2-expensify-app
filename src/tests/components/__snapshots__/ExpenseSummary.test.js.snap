// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h3
      className="page-header__title"
    >
      You are viewimg 
      <span>
        1
      </span>
       
      expense
       whose total amount is 
      <span>
        $2.35
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h3
      className="page-header__title"
    >
      You are viewimg 
      <span>
        23
      </span>
       
      expenses
       whose total amount is 
      <span>
        $2,354,689,866.70
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
